Bug 936647 - [Messages][Drafts] Preview drafts threads in thread list

rwaldron opened the pull request:
https://bugzilla.mozilla.org/show_bug.cgi?id=936647
- index.html: 
  - messages-thread-tmpl
    - Adds "icon-draft" class to "icon icon-read" classes
    - Adds data-l10n-id="no-receipient" 
    - Adds data-list=draft|thread
- js/desktop-only/mobilemessage.js
  - Updates Draft bootstrapping data
- js/drafts.js
  - Adds read-once and cache to Drafts.load
  - Adds unique id mechanism
  - Updates Drafts.delete to allow deleting all drafts for a threadId
  - Renames Drafts.byId => Drafts.byThreadId
  - Adds Drafts.get(draftId)
  - Updates Drafts.load to abide a local isCached flag.
    - Only load from storage when isCached is false.
  - Updates Drafts.load to revive stored Draft objects as actual Draft objects
  - Adds Draft.List.prototype.latest that points at the latest draft in a Draft.List (ie. for a specified threadId or null key)
  - Adds function guid for creating quasi-unique ids
- js/message_manager.js
  - Adds `draft` property to hold draft object to coordinate with New Message view
  - Calls Drafts.load() from init
  - Calls ThreadListUI.renderDrafts() from onHashChange when hash = #thread-list
- js/thread_list_ui.js
  - Wraps contents in IIFE
  - Adds draftLinks property to map DOM nodes to draftId
  - Updates setContact to include lookup from draft
  - Updates handleEvent to capture draft object on click. Stops looking at 'threads-container'
  - Updates delete to create two lists of deletable items from the selectedInputs array. List is determined by data-list property value.
    - Drafts are deleted similarly to Threads, but will exit immediately if no Threads are to be deleted
    - Drafts.store() is called upon completing delete operation
  - Adds ThreadListUI.renderDrafts to create "Thread List Items" from Draft objects that do not have their own thread 
  - Calls ThreadListUI.renderDrafts from ThreadListUI.renderThreads
  - Updates ThreadListUI.createThread to create a "Thread List Item" from a Draft object, or to mark a Thread that has a draft.
  - Updates rendered markup for "Thread List Item"
  - Updates timestamp handling in ThreadListUI.appendThread 
- js/thread_ui.js
  - Updates ThreadUI.saveMessageDraft new or replacement drafts
- js/threads.js
  - Expose Thread constructor
  - Adds Thread.prototype.drafts accessor to Drafts.byThreadId(this.id);
  - Adds Thread.prototype.hasDrafts accessor as short hand
- Adds tests to: 
  - test/unit/compose_test.js
  - test/unit/drafts_test.js
  - test/unit/media/pixel.jpg
  - test/unit/message_manager_test.js
  - test/unit/mock_async_storage.js
  - test/unit/mock_drafts.js
  - test/unit/mock_navigatormoz_sms.js
  - test/unit/thread_list_ui_test.js
  - test/unit/thread_ui_test.js
  - test/unit/threads_test.js
- Passing both lint and hint

rwaldron committed:
Bug 935177 - [Messages][Drafts] Update "discard message" confirmation prompt to options menu/dialog; add "save draft" no-op r=julienw

https://bugzilla.mozilla.org/show_bug.cgi?id=935177

- New strings:
  - save-as-draft      = Save as Draft
  - discard-message = Discard
- Removes window.confirm mechanism
- Implements Multi-option prompt with OptionMenu
- Basic tests to assert the correct operations are performed by the option handlers

Signed-off-by: Rick Waldron <waldron.rick@gmail.com>

evhan55 committed:
Bug 936641 - [Messages][Drafts] Create message drafts upon user interaction

evhan55 commented on apps/sms/js/drafts.js:
```diff
@@ -71,15 +72,16 @@
         // If the new draft is distinct from the stored one
         // then push and save a new draft
         if (isDistinct(stored, draft)) {
           thread.push(draft);
```
So, I had a question.  Did we decide to only ever store one draft per threadId?
If so, shouldn't Drafts.add remove the old draft at that threadId if it exists?
This has never been written into the add, and I wonder if we want to change it to do that.

Is there any case where we want to store old drafts for a single thread?

gnarf replied:
The "no recipient" thread maybe?

rwaldron replied:
@evhan55 This makes sense.

evhan55 commented on apps/sms/js/drafts.js:
```diff
@@ -188,6 +265,15 @@
       return priv.get(this).length;
     },
     /**
+     * latest
```
If we change it to always keep one draft per threadId, would something like this be necessary?

gnarf replied:
Would `last` make more sense?

rwaldron replied:
@evhan55 yes, because `byThreadId` returns a `Draft.List` and sometimes we just want the `Draft` object

@gnarf I had `last` originally, but kept thinking of it as `latest`, so I changed it.

evhan55 commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -287,6 +342,51 @@ var ThreadListUI = {
+          recipients = draft.recipients.length ?
+            draft.recipients : [''];
+          body = draft.content.find(function(content) {
+            if (typeof content === 'string') {
```
Will this also find blobs?

rwaldron replied:
Aren't these stored as plain objects?

julienw replied:
I will trust you here, I don't follow everything

gnarf commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -2,12 +2,16 @@
+(function(exports) {
 'use strict';
+var noRecipient = navigator.mozL10n ?
+  navigator.mozL10n.get('noRecipient') : 'no recipient';
```
This is going to be a problem -- If the language is changed while the app is open, this value will never be updated?

rwaldron replied:
Updated in next push

gnarf commented on apps/sms/js/drafts.js:
```diff
@@ -216,6 +302,12 @@
+      throw new Error('Draft id must be a number');
+    }
+
+    this.id = draft.id || +(Date.now() + '' + (Math.random() * 1000 | 0));
```
Is `+(Date.now() + '' + (Math.random() * 1000 | 0))` meant to be a unique id?

Can we make it this:

``` js
function getUID() {
  return +(Date.now() + '' + (Math.random() * 1000 | 0));
}
// .....
this.id = draft.id || getUID();
```

Feels like a readability improvement to me.

rwaldron replied:
Updated in next push

evhan55 commented on apps/sms/js/thread_ui.js:
```diff
@@ -2510,6 +2510,13 @@ var ThreadUI = global.ThreadUI = {
+    if (threadId) {
+      li = document.getElementById('thread-' + threadId);
+      li.classList.add('draft');
+      ThreadListUI.draftLink.set(li, draft.id);
```
draftLink => draftLinks

gnarf commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -287,6 +342,51 @@ var ThreadListUI = {
+            isDraft: true
+          });
+        }
+      }.bind(this));
```
can we just pass a thisArg here instead of binding?

Protonk replied:
thread lists don't allow that right now. I think this might be a case for my proposal to just use forEach:

```
DraftList.prototype.forEach = function(callback, scope) {
  var arr = Array.prototype.slice.call(priv.get(this), 0);
  Array.prototype.forEach.call(arr, callback, scope);
};
```

rwaldron replied:
~~Just do this in your patch~~

_edit_
There was a reason we did this, which was to _not_ leak out the intentionally encapsulated drafts array.

rwaldron replied:
@gnarf 

> can we just pass a thisArg here instead of binding?

https://github.com/rwaldron/gaia/blob/v1.3-drafts/apps/sms/js/drafts.js#L174-L206

This is to prevent future additions from trying to `push` or `pop` and expect that to work and changes to persist.

gnarf replied:
Fair enough - maybe we should use `each` instead of `forEach` because I assumed this was a real array :)

rwaldron replied:
fine with me

rwaldron committed:
Bug 936647 - [Messages][Drafts] Preview drafts threads in thread list r=julienw

https://bugzilla.mozilla.org/show_bug.cgi?id=936647

- index.html:
  - messages-thread-tmpl
    - Adds "icon-draft" class to "icon icon-read" classes
    - Adds data-l10n-id="no-receipient"
    - Adds data-list=draft|thread

- js/desktop-only/mobilemessage.js
  - Updates Draft bootstrapping data

- js/drafts.js
  - Adds read-once and cache to Drafts.request
  - Adds unique id mechanism
  - Updates Drafts.delete to allow deleting all drafts for a threadId
  - Renames Drafts.byId => Drafts.byThreadId
  - Adds Drafts.get(draftId)
  - Updates Drafts.request to abide a local isCached flag.
    - Only load from storage when isCached is false.
  - Updates Drafts.request to revive stored Draft objects as actual Draft objects
  - Adds Draft.List.prototype.latest that points at the latest draft in a Draft.List (ie. for a specified threadId or null key)
  - Adds function guid for creating quasi-unique ids

- js/message_manager.js
  - Adds `draft` property to hold draft object to coordinate with New Message view
  - Calls Drafts.request() from init
  - Calls ThreadListUI.renderDrafts() from onHashChange when hash = #thread-list

- js/thread_list_ui.js
  - Wraps contents in IIFE
  - Adds draftLinks property to map DOM nodes to draftId
  - Updates setContact to include lookup from draft
  - Updates handleEvent to capture draft object on click. Stops looking at 'threads-container'
  - Updates delete to create two lists of deletable items from the selectedInputs array. List is determined by data-list property value.
    - Drafts are deleted similarly to Threads, but will exit immediately if no Threads are to be deleted
    - Drafts.store() is called upon completing delete operation
  - Adds ThreadListUI.renderDrafts to create "Thread List Items" from Draft objects that do not have their own thread
  - Calls ThreadListUI.renderDrafts from ThreadListUI.renderThreads
  - Updates ThreadListUI.createThread to create a "Thread List Item" from a Draft object, or to mark a Thread that has a draft.
  - Updates rendered markup for "Thread List Item"
  - Updates timestamp handling in ThreadListUI.appendThread

- js/thread_ui.js
  - Updates ThreadUI.saveMessageDraft new or replacement drafts

- js/threads.js
  - Expose Thread constructor
  - Adds Thread.prototype.drafts accessor to Drafts.byThreadId(this.id);
  - Adds Thread.prototype.hasDrafts accessor as short hand

- Adds tests to:
  - test/unit/compose_test.js
  - test/unit/drafts_test.js
  - test/unit/media/pixel.jpg
  - test/unit/message_manager_test.js
  - test/unit/mock_async_storage.js
  - test/unit/mock_drafts.js
  - test/unit/mock_navigatormoz_sms.js
  - test/unit/thread_list_ui_test.js
  - test/unit/thread_ui_test.js
  - test/unit/threads_test.js

- Passing both lint and hint

Signed-off-by: Rick Waldron <waldron.rick@gmail.com>

julienw commented:
lots of jshint errors

julienw commented:
gjslint issues as well btw ;)

julienw commented on apps/sms/index.html:
```diff
@@ -311,15 +311,15 @@ <h1 class="subject">
           <img src="">
         </aside>
-        <p class="name">${number}</p>
+        <p class="name" data-l10n-id="no-receipient">${number}</p>
```
nit: recipient

rwaldron replied:
fixed

rwaldron commented:
@julienw both the lint and hint errors were fixed probably at the same moment you wrote those :)

julienw commented on apps/sms/style/sms.css:
```diff
@@ -139,6 +139,14 @@ form[role="dialog"][data-type="edit"] header menu[type="toolbar"] button {
+}
+
+[data-type="list"] .draft aside.icon-draft {
+  background: url("../style/images/draft.png") no-repeat left center / 1rem;
```
3 comments on this:
- can we factorize with icon-unread? Have an "icon" class for example? Only the background-image changes between states.
- can you investigate if we can have a .draft class on the thread node itself? And .unread too? Maybe even make the icon a generated content (although I'm not sure of this one, maybe it's overkill) ?
- I think we can have both unread and draft, and your current markup does not allow this.

rwaldron replied:
> can we factorize with icon-unread? Have an "icon" class for example? Only the background-image changes between states.

We could, but there is already a class called .icon. Maybe a more thorough cleanup should be done, but in a different patch?

> can you investigate if we can have a .draft class on the thread node itself? And .unread too? Maybe even make the icon a generated content (although I'm not sure of this one, maybe it's overkill) ?

We do, for setting the opacity to 0.5

> I think we can have both unread and draft, and your current markup does not allow this.

I'm not sure how that's possible (visually)—the two icons exist at the exact same place :(

julienw replied:
ok, I get it, we can have basically these states:
- a "pure draft", that is a "new message draft", has is semi transparent and has the pencil icon
- a "thread draft" has the pencil icon but is fully opaque

Now, what should we have for threads that have a draft but there is an unread message in it? IINM This means the most recent message in the thread is more recent than when the draft was created (because when the draft is created, all messages are read)

This is both a UX+Visual Design issue, let me raise this on the bug

julienw replied:
that said, I'm open with moving forward and fixing whatever nees to be fixed here in a follow-up

rwaldron replied:
> that said, I'm open with moving forward and fixing whatever nees to be fixed here in a follow-up

Ok, good to know and I agree

julienw commented on apps/sms/js/drafts.js:
```diff
@@ -63,23 +64,32 @@
         // If the new draft is distinct from the stored one
         // then push and save a new draft
-        if (isDistinct(stored, draft)) {
+        if (canSkipDistinction || isDistinct(stored, draft)) {
```
it seems like we'll always skip distinction if there is a stored draft? Is there a case where isDistinct is used here?

rwaldron replied:
Two cases:
- `canSkipDistinction` - this is a draft for an existing thread (so we just get rid of the old one and replace it with this new one)
- `isDistinct(stored, draft)` - this is used to check if this new draft is different from the last saved draft for drafts that have a `null` threadId (new message drafts)

julienw replied:
if we don't have a threadId, stored will be undefined here (`thread[thread.length - 1]`). So isDistinct will always return true. Is it what you want?

julienw replied:
about the logic: this seems great when the user presses the "new message" button, write something etc. But when the user loads an existing draft, I think that saving that draft should update the existing draft (or remove it/add a new one). But I don't see you're using draft.id here.

rwaldron replied:
> if we don't have a threadId, stored will be undefined here (thread[thread.length - 1]). So isDistinct will always return true. Is it what you want?

There will always be a "threadId", either a number that corresponds to a thread conversation or `null` to represent new message drafts that have no recipients or cannot be matched to an existing conversation. `null` is actually the case that will have _many_ drafts objects associated with it.

>  But when the user loads an existing draft, I think that saving that draft should update the existing draft (or remove it/add a new one). 
- This is exactly what happens when the draft is for an existing conversation thread
- For `null` thread (ie. new message drafts) we update the last saved draft if nothing has changed about the content or recipients

julienw commented on apps/sms/js/drafts.js:
```diff
@@ -95,13 +105,25 @@
+          if (thread && typeof draft.id === 'undefined') {
+            thread.length = 0;
+          } else {
+            thread.forEach(function(stored, i) {
```
will fail if thread is falsy

rwaldron replied:
fixed

julienw commented on apps/sms/js/drafts.js:
```diff
@@ -95,13 +105,25 @@
+          } else {
+            thread.forEach(function(stored, i) {
+              if (stored.id === draft.id) {
+                index = i;
```
you may want to bail out of this loop here?

In that case, use `some` instead of `forEach`, and return true?

rwaldron replied:
fixed

julienw commented on apps/sms/js/drafts.js:
```diff
@@ -95,13 +105,25 @@
+        thread = draftIndex.get(draft.threadId);
         index = thread.indexOf(draft);
+
+        if (index === -1) {
```
nit: this algorithm could use a comment

rwaldron replied:
done

julienw commented on apps/sms/js/drafts.js:
```diff
@@ -117,10 +139,35 @@
+          draft = records.find(function(record) {
+            // Ensure a number is used for the comparison,
+            // as this value may come from a dataset property.
+            // Curse the day that state was stored in the DOM.
```
we don't need this last sentence ;)

rwaldron replied:
hehe... forgot about that 

fixed

julienw commented on apps/sms/js/drafts.js:
```diff
@@ -151,14 +198,61 @@
+        setTimeout(function() {
+          handler();
+        });
+        return;
```
this return is useless

rwaldron replied:
Fixed

julienw replied:
I still see it ;)

rwaldron replied:
Indeed. This will be "actually fixed" in the next push ;)

julienw commented on apps/sms/js/drafts.js:
```diff
@@ -151,14 +198,61 @@
+              //
+              // Once ES6 syntax is allowed,
+              // replace the map operations below with the following line:
+              // record[1].map(draft => new Draft(draft));
```
this seems quite obvious ;) this comment about ES6 syntax is unnecessary, please remove it

rwaldron replied:
We have several lines like this already and it's good to keep track. Either way, the updated code won't use map, so it's being removed anyway.

julienw commented on apps/sms/js/drafts.js:
```diff
@@ -151,14 +198,61 @@
+              //            with that threadId or null key
+              //
+              // Replace each plain object in record[1] with a
+              // real draft object.
```
nit: please enclose this comment by /\* */ instead of these //

rwaldron replied:
fixed

julienw commented on apps/sms/js/drafts.js:
```diff
@@ -151,14 +198,61 @@
+              //
+              //
+              record[1] = record[1].map(function(draft) {
+                return new Draft(draft);
```
you could add a `Draft.fromJSON` constructor method, and then it would just be: `record[1].map(Draft.fromJSON)`. ;)

rwaldron replied:
I had considered something similar, but this data isn't serialized JSON, so "fromJSON" isn't accurate. The whole thing has been replaced though, so this line is irrelevant now

julienw commented on apps/sms/js/drafts.js:
```diff
@@ -151,14 +198,61 @@
+          // Revive as Draft objects by constructing new Draft from
+          // each plain object returned from storage.
+          if (records !== null) {
+            records = records.map(function(record) {
```
here we're creating a new object. Since it could be quite large, I wonder if we could not change directly `records` instead.

julienw replied:
then instead of allocating the whole array before freeing the old one, it would free the individual items once at a time.

rwaldron replied:
> , I wonder if we could not change directly records instead.

Updating to this: 

``` js
if (records !== null) {
  /*
    record[0] is the threadId or null key
    record[1] is the array of draft objects associated
               with that threadId or null key

    Replace each plain object in record[1] with a
    real draft object.

  */
  rlength = records.length;

  for (var i = 0; i < rlength; i++) {
    drafts = records[i][1];
    dlength = drafts.length;

    for (var j = 0; j < dlength; j++) {
      drafts[j] = new Draft(drafts[j]);
    }
  }
}
```

julienw replied:
:+1:

julienw commented on apps/sms/js/drafts.js:
```diff
@@ -224,6 +333,9 @@
 
-  exports.Draft = Draft;
+  function guid() {
+    return +(Date.now() + '' + (Math.random() * 1000 | 0));
```
tell me if it's stupid: how about just incrementing a counter that you'll store in DB?

rwaldron replied:
I don't think it's stupid, but I didn't want to rely too heavily on the DB

julienw replied:
oki

julienw commented on apps/sms/js/message_manager.js:
```diff
@@ -266,6 +269,8 @@ var MessageManager = {
         var optionsButton = document.getElementById('messages-options-icon');
         optionsButton.parentNode.appendChild(optionsButton);
 
+        ThreadListUI.renderDrafts();
```
is it possible that we haven't finished loading when we reach this?

rwaldron replied:
`renderDrafts` does a `Drafts.load`

julienw replied:
great

rwaldron replied:
which is now `Drafts.request`

julienw commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -2,12 +2,14 @@
 'use strict';
 
 var ThreadListUI = {
+  draftLinks: new Map(),
```
for unit tests sake, can we initialize it in init() ?

you can keep the property definition here though.

rwaldron replied:
Fixed

julienw replied:
should this be a WeakMap?

julienw commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -91,21 +93,23 @@ var ThreadListUI = {
+      number = thread.participants[0];
+      others = thread.participants.length - 1;
+    } else {
+      if (draft) {
```
nit: `else if (draft) {`

rwaldron replied:
fixed

julienw commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -91,21 +93,23 @@ var ThreadListUI = {
+        number = draft.recipients[0];
+        others = draft.recipients.length - 1;
+      }
     }
```
Hopefully we'll have a Thread abstraction that will handle this one day (`Thread.fromDraft` and `Thread.fromMessageThread`)

julienw replied:
Another idea that is doable right now: maybe the Draft object can have a `asThread` method that would just return a thread-like json?

julienw replied:
another idea is to have drafts expose `participants` instead of `recipients`.

rwaldron replied:
> Hopefully we'll have a Thread abstraction that will handle this one day (Thread.fromDraft and Thread.fromMessageThread)

Yes, I think I own that ticket

>  Draft object can have a asThread method that would just return a thread-like json?

Do you mean object—there is no JSON. Drafts are always objects, ie. never appear serialized in the application

> another idea is to have drafts expose participants instead of recipients.

But they're not participants yet and it would create the same mismatch for the "send" case, which are "recipients".

julienw replied:
> Do you mean object—there is no JSON. Drafts are always objects, ie. never appear serialized in the application

Yep, I meant a plain object (as opposed to a typed object)

> But they're not participants yet and it would create the same mismatch for the "send" case, which are "recipients".

They're future participants ;-)
I don't get the mismatch part of your sentence.

julienw commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -287,6 +340,51 @@ var ThreadListUI = {
+    // Load and render all thread-less drafts. Drafts.load only
+    // makes an asyncStorage request (IndexedDB) the first time it's called.
+    // Each subsequent time, the threadless drafts come directly
+    // from the the in-memory draft index.
```
maybe the name "load" is not well-chosen then. You could keep `load` internal (or even expose it, it's quite meaningfull when we call it in `init`) and have `get` from the outside? This could even be aliases, I don't really mind.

evhan55 replied:
I had the same reaction at first to 'load' being called here in renderDrafts, but I don't mind either way.

rwaldron replied:
Fixed. `Drafts.request` is always a request for drafts, regardless of where they come from

julienw commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -287,6 +340,51 @@ var ThreadListUI = {
+
+        // If there is currently no list item rendered for this
+        // draft, then proceed.
+        if (draftIds.indexOf(draft.id) === -1) {
```
can't we use `this.draftLinks.has(draft.id)` ? And then you can get rid of the `draftIds` array

rwaldron replied:
The keys in `this.draftLinks` are the actual DOM node (`li`) that represents the draft object in the UI—note that:

``` js
this.draftLinks.forEach(function(v, k) {
  draftIds.push(v);
});
```

is capturing a list of values, not the keys

julienw replied:
right

I don't like this algorithm much:
- first we iterate over the draftLinks array
- then for each threadless draft we call indexOf (which has linear complexity)

I think the draftLinks map should be reversed: the key should be the id, the value the node. There is only one place where you need to get the id from the node and it happens from one node only; on the contrary, here, you're doing it for all threads. So I think the key should really be the id instead.

The alternative could be to build a more specialized object where the ids could be stored as an object key allowing a quick "has" operation.

julienw replied:
>  There is only one place where you need to get the id from the node and it happens from one node only

I missed it's happening in a loop. Probably that loop (in the handleEvent method) could be rewritten to check for a class to find the correct node, and only then look into the map. But the general comment still applies.

julienw commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -242,16 +264,47 @@ var ThreadListUI = {
+          // be derived and passed on to delete() (these don't have
+          // an associated thread). In the former case,
+          // pass on a threadId so that delete can simply eliminate all
+          // drafts for the selected thread.
```
I'd say the "thread" case won't happen in this loop, since you're looping `list.drafts`? Am I missing something?

rwaldron replied:
> I'd say the "thread" case won't happen in this loop, since you're looping list.drafts? Am I missing something?

Yes, if I "select all" or even "select one" thread item (that has a draft associated with it) from thread list ui, then I want to delete the draft associated with the thread as well. The "select all" case is most illustrative: if I select all and delete, it's strange to see a draft still appearing in the list.

julienw replied:
now I get it, I mixed the dataset "mode" and the dataset "list".

So while I won't block on this, here is how I would have done the delete case:
- find all selected new message drafts and delete them (part of your current algorithm, except you know they are new message drafts only
- for all threads
  *\* delete associated drafts (very easy with the API you already have)
  *\* delete messages (current code)

Advantages:
- we use only the "mode" dataset (and we get rid of the "list" dataset)
- it's a more linear algorithm

rwaldron replied:
fixed.

julienw commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -242,16 +264,47 @@ var ThreadListUI = {
+          this.removeThread(id);
+        }, this);
+
+        Drafts.store();
```
couldn't all or part of this be handled by the Drafts object?

rwaldron replied:
We specifically avoided calling `Drafts.store()` from `Drafts.add()` and `Drafts.delete()` to reduce DB writing to a few explicit calls

julienw commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -287,6 +340,51 @@ var ThreadListUI = {
+    // makes an asyncStorage request (IndexedDB) the first time it's called.
+    // Each subsequent time, the threadless drafts come directly
+    // from the the in-memory draft index.
+    Drafts.load(function(threadless) {
```
do we get thread-bound drafts here?

rwaldron replied:
No, for a two reasons: 
1. The application doesn't need thread-bound drafts in `ThreadListUI.renderDrafts` (those are all "isDraft")
2. Thread list items that are actually a thread only need to know if they are a "hasDraft"
3. Considering 1 & 2, it makes sense to provide only the data needed instead of providing all of the data and forcing the consumer to filter

julienw commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -287,6 +340,51 @@ var ThreadListUI = {
+            unreadCount: 0,
+            lastMessageType: draft.type,
+            isDraft: true
+          });
```
a `Draft.asThread` could be useful here as well (and would do basically this)

rwaldron replied:
If drafts.js needs to construct Thread objects, it creates a circular dependency because threads.js already depends on drafts.js for Thread.prototype.drafts and Thread.prototype.hasDrafts

julienw replied:
In the Java World, I used to use the concept of "value objects", that is simple objects that only stores data. That's how the Thread and Draft objects should be IMO.

But here, what I suggested was really a `asThread` method returning a plain object, exactly like the one you pass as parameter to appendThread:

``` javascript
Draft.prototype.asThread = function() {
  var recipients = this.recipients.length ? this.recipients : [''];
  var body = this.content.find(function(content) {
    if (typeof content === 'string') {
      return true;
    }
  });

  return {
    id: this.id,
    participants: recipients,
    body: body || '',
    timestamp: this.timestamp,
    unreadCount: 0,
    lastMessageType: this.type,
    isDraft: true
  };
}
```

julienw replied:
and then in renderDrafts:

``` javascript
if (draftIds.indexOf(draft.id) === -1) {
  ThreadListUI.appendThread(draft.asThread());
}
```

Don't you think that would be so elegant? :)

julienw commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -360,26 +460,63 @@ var ThreadListUI = {
     var li = document.createElement('li');
-    var timestamp = thread.timestamp.getTime();
+    var timestamp = typeof thread.timestamp !== 'number' ?
+      thread.timestamp.getTime() : thread.timestamp;
```
use `+thread.timestamp` ?

julienw commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -360,26 +460,63 @@ var ThreadListUI = {
+      }
+
+      ThreadListUI.draftLinks.set(
+        li, hasDrafts ? draft.id : thread.id
```
I think this is wrong here: you don't want the draft id if it's a thread-bound draft, if I understand correctly the other code where you use draftLinks.

But maybe I don't follow it correctly.

julienw commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -360,26 +460,63 @@ var ThreadListUI = {
 
     // Render markup with thread data
     li.innerHTML = this.tmpl.thread.interpolate({
+      hash: isDraft ? '#new' : '#thread=' + id,
```
how about having a `#new&draft=<draft_id>` ?

julienw replied:
would need to replace all `hash === '#new'` occurences by startsWith...

rwaldron replied:
I wanted that at first, but that would require changing all of the logic the depends on checks like `location.hash === '#new'` or using `location.hash` in a `switch` with case matching `#new`. (This sounds like I'm repeating back to you, but I'm just trying to be specific about the existing cases)

julienw replied:
yeah

I need to finish the navigation stuff grrrmbl

julienw commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -441,7 +584,15 @@ var ThreadListUI = {
   appendThread: function thlui_appendThread(thread) {
-    var timestamp = thread.timestamp.getTime();
+    var timestamp = typeof thread.timestamp !== 'number' ?
+      thread.timestamp.getTime() : thread.timestamp;
```
`+thread.timestamp`?

rwaldron replied:
Fixed

julienw commented on apps/sms/js/thread_ui.js:
```diff
@@ -2500,16 +2499,29 @@ var ThreadUI = global.ThreadUI = {
 
+    // If an existing thread list item is associated with
+    // the presently saved draft, remove the thread and
+    // re-render to adtivate the new draft-preview
```
nit: activate

rwaldron replied:
fixed

julienw commented on apps/sms/js/thread_ui.js:
```diff
@@ -2500,16 +2499,29 @@ var ThreadUI = global.ThreadUI = {
+      // the drafts timestamp.
+      thread.timestamp = draft.timestamp;
+
+      ThreadListUI.appendThread(thread);
```
can you look if `ThreadListUI.updateThread` is usable?

rwaldron replied:
Fixed. 

This required implementing an updated `Thread.create` (ie. rewriting `Threads.createThreadMockup` since `ThreadListUI.updateThread` was expected to call through to that function for creating thread objects from messages

julienw commented on apps/sms/test/unit/drafts_test.js:
```diff
@@ -6,10 +6,25 @@ requireApp('sms/js/is-equal.js');
+      value: MockasyncStorage,
+      configurable: true,
+      writable: true
+    });
```
won't mocksHelper work flawlessly for this?

julienw commented on apps/sms/test/unit/drafts_test.js:
```diff
@@ -68,13 +83,19 @@ suite('Drafts', function() {
+  suiteTeardown(function() {
+    asyncStorage.getItem = realASgetItem;
+    asyncStorage.setItem = realASsetItem;
+    asyncStorage.removeItem = realASremoveItem;
```
just use mocksHelper

rwaldron replied:
Fixed

julienw commented on apps/sms/test/unit/mock_async_storage.js:
```diff
@@ -0,0 +1,19 @@
+  getItem: function(key, callback) {
+    setTimeout(function() {
+      callback(MockasyncStorage.data);
+    });
```
be careful, this will give intermittent failures.

I think the best way to do this is to have an empty function, and use sinon to "yield" the desired value only in the tests where you need it.

rwaldron replied:
Fixed

rwaldron commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -388,6 +525,12 @@ var ThreadListUI = {
+      );
+    }
+
     return li;
```
I wish... I had this using templates, but then @gnarf educated me re: `navigator.mozL10n.localize` so I had to change it to do post-template-rendering operation :(

julienw commented on apps/sms/js/drafts.js:
```diff
@@ -63,23 +64,32 @@
+        if (canSkipDistinction || isDistinct(stored, draft)) {
           thread.push(draft);
-          draftIndex.set(id, thread);
+          draftIndex.set(threadId, thread);
```
if threadId is null, is it correct?

rwaldron replied:
It could be either `null` or a number (the former for new message drafts and the latter for existing thread drafts)

julienw replied:
ok, could you add a comment explaining that a null ThreadId will yield all thread-less drafts?

julienw commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -91,21 +96,21 @@ var ThreadListUI = {
+    } else if (draft) {
+      number = draft.recipients[0];
+      others = draft.recipients.length - 1;
     }
```
maybe we should at least add a console.error call in a `else` block here

rwaldron commented on apps/sms/js/drafts.js:
```diff
@@ -18,7 +19,7 @@
    */
   function isDistinct(a, b) {
```
Maybe, but equal isn't quite right, since the timestamp of the draft objects will never be equal.

julienw commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -242,16 +265,47 @@ var ThreadListUI = {
-        return input.value;
-      });
+      list = this.selectedInputs.reduce(function(list, input) {
+        list[input.dataset.list].push(+input.value);
```
`input.dataset.mode`

julienw replied:
for get it, I see that we use both mode and list.

Is it really useful to have both of them?

julienw replied:
especially with different values...

julienw replied:
as a general note, when reading the code, I find it quite difficult to understand when we deal with only new-message-drafts and when we deal with both draft types, and I think we can do better.

rwaldron replied:
> Is it really useful to have both of them?

No, `mode` is left over and needs to be removed.

>  I find it quite difficult to understand when we deal with only new-message-drafts and when we deal with both draft types, and I think we can do better.

I followed the approach you described here: https://bugzilla.mozilla.org/show_bug.cgi?id=919971#c16
- new message drafts are rendered in renderDrafts, which is called by renderThreads
- existing conversation drafts are just a bit of extra data added to a regular thread list item

rwaldron commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -67,6 +70,8 @@ var ThreadListUI = {
       this.updateContactsInfo.bind(this)
     );
+
+    this.draftLinks = new Map();
```
No, because we need to be able to iterate over the stored key/value pairs.

I'm glad you mentioned this, because I need to remove these when they are removed from the DOM

julienw commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -287,6 +341,34 @@ var ThreadListUI = {
+    });
+
+    // Request and render all thread-less drafts.
+    Drafts.request(function() {
```
> No, for a two reasons:
> 
>   The application doesn't need thread-bound drafts in ThreadListUI.renderDrafts (those are all "isDraft")
>    Thread list items that are actually a thread only need to know if they are a "hasDraft"
>    Considering 1 & 2, it makes sense to provide only the data needed instead of providing all of the data and forcing the consumer to filter

I fully agree, it's just that the names Drafts.load or Drafts.request are not very self-explanatory. How about `Drafts.ThreadLess.request` ? Or `Drafts.requestThreadLess` ? Or `Drafts.request({ type: 'threadless' }, function() {...})`?

I don't like the last one because it nearly forces you to make it work for all types, but how about the first ones? Or anything else that is more self-explanatory?

julienw commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -357,29 +441,70 @@ var ThreadListUI = {
-    var timestamp = thread.timestamp.getTime();
-    var lastMessageType = thread.lastMessageType;
-    var participants = thread.participants;
+    var timestamp = record.timestamp.getTime();
```
`var timestamp = +record.timestamp` ?

julienw commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -409,21 +540,22 @@ var ThreadListUI = {
-    if (existingThreadElement &&
-      existingThreadElement.dataset.time > timestamp) {
-      // If the received SMS it's older that the latest one
+    var timestamp = thread.timestamp.getTime();
```
`var timestamp = +thread.timestamp` ?

julienw commented on apps/sms/js/threads.js:
```diff
@@ -18,47 +20,104 @@
-      if (message.delivery === 'received' ||
-          message.delivery === 'not-downloaded') {
-        participants = [message.sender];
+  Thread.create = function(record, options) {
```
there are a lot of cases in this function.

I really think having separate constructors `Thread.fromMessage`, `Thread.fromDraft` would be a lot more readable in the end. And probably more efficient (more linear code, less if blocks).

julienw replied:
and you would not need to "guess" the specific cases then, hence less possible difficult-to-debug errors

julienw commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -357,29 +441,70 @@ var ThreadListUI = {
+
+      ThreadListUI.draftLinks.set(
+        li, hasDrafts ? draft.id : record.id
+      );
```
Still this:

I think this is wrong here: you don't want the draft id if it's a thread-bound draft, if I understand correctly the other code where you use draftLinks.

julienw commented on apps/sms/js/is-equal.js:
```diff
@@ -0,0 +1,282 @@
+// THIS FILE IS NOT SUBJECT TO CHANGE
```
can you move this to a js/ext/ subdirectory?

and add ext/\* to .jshintignore ?

(that's what I did in my WIP branch with the promise library)

julienw commented on apps/sms/js/drafts.js:
```diff
@@ -1,7 +1,42 @@
+        }
+      }
+      // else check whether content or subject match
+      return !isEqual(a.content, b.content) ||
```
is there no easier way than importing an external library to check this?

julienw commented on apps/sms/test/unit/drafts_test.js:
```diff
@@ -251,77 +345,120 @@ suite('Drafts', function() {
+      // d6 is almost the same as d5, b/w different content
+      Drafts.add(d6);
+
+      assert.isTrue(spy.calledTwice);
```
sinon.assert.calledTwice(spy)

julienw commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -388,6 +513,12 @@ var ThreadListUI = {
+      navigator.mozL10n.localize(
+        li.querySelector('.name'), 'no-recipient'
+      );
+    }
```
nit: you should do this in setContact, in the `if (!number)` block

gnarf replied:
counter nit: It's not possible to actually do that in the above block, unless we also allow data-l10n-id injection in the template.  Otherwise if language is switched while this `(no recipient)` is rendered, it won't change language.

I told rick to do it this way because of that giant l10n bug I dealt with all of the edge cases,  If you ever catch yourself calling `navigator.mozL10n.get` you're probably doing it wrong.

julienw replied:
I think you misunderstood me. I didn't ask to do this in the above block, but in the block at https://github.com/rwaldron/gaia/pull/1/files#diff-e2424773c46c9d5f4ff27c9c20b1bb0dL103

But now I realize that no-recipient is the default data l10n id in the template (see https://github.com/rwaldron/gaia/pull/1/files#diff-e2424773c46c9d5f4ff27c9c20b1bb0dL103). Maybe this is wrong (this would make the thread show 'no recipient' while it tries to match contacts with numbers when the thread list is loading) ?

rwaldron replied:
Fixed

julienw commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -91,21 +98,27 @@ var ThreadListUI = {
 
   setContact: function thlui_setContact(node) {
```
nit: can you add an early error + return in case node.dataset.threadId is falsy?

julienw commented on apps/sms/js/drafts.js:
```diff
@@ -95,13 +105,37 @@
+          if (thread && typeof draft.id === 'undefined') {
+            thread.length = 0;
+          } else {
+            if (thread) {
```
nit: else if (thread)

julienw commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -201,11 +221,19 @@ var ThreadListUI = {
+
+    if (li) {
+      parent = li.parentNode;
+      parent.removeChild(li);
```
`li.remove();`

julienw replied:
mmm we need the parent later, ok, let's keep it like it is.

gnarf replied:
That's a thing?  Not according to DOM Spec as far as I can tell... Always done `N.parentNode.removeChild(N)` myself, never heard of this method :)

julienw replied:
http://dom.spec.whatwg.org/#interface-childnode ;) Enjoy!

rwaldron replied:
yeah... I'm cool with that, but I was just copying existing code. The last time I tried to use .remove, it wasn't actually available in the gecko on the device.

julienw replied:
yeah and it won't simplify much here since we need the parent later.

rwaldron replied:
Looks good now.

julienw commented on apps/sms/js/thread_list_ui.js:
```diff
@@ -287,6 +341,28 @@ var ThreadListUI = {
+        // draft, then proceed.
+        if (!this.draftRegistry[draft.id]) {
+          this.appendThread(
+            Thread.create(draft)
```
you could use Thread.fromDraft here, if you feel like it

julienw commented on apps/sms/js/thread_ui.js:
```diff
@@ -2500,16 +2499,26 @@ var ThreadUI = global.ThreadUI = {
+
+      // Overwrite the thread's own timestamp with
+      // the drafts timestamp.
+      thread.timestamp = draft.timestamp;
```
some thoughts for later: I wonder if this should not be done in Drafts.add.

also, I wonder how we can do this at init time

rwaldron replied:
`Drafts.add` shouldn't do _anything_ to Threads, especially if it involves `Threads.get(threadId)` because that creates a circular dependency between Drafts and Threads

julienw commented on apps/sms/js/message_manager.js:
```diff
@@ -42,6 +42,8 @@ var MessageManager = {
       callback();
     }
+
+    Drafts.request();
```
wondering if this is really useful.

If we do renderDrafts in the same time and this is not finished, we'll end up doing twice the same request, which is probably not a good thing.

So 2 possibilities:
- either we do something smarter in Drafts.request: if a request is ongoing, just store the handler, and run it when the ongoing request is finished
- or we remove this line now and we file a bug about doing this properly later

rwaldron replied:
According to @evhan55, it fixes a bug that was found, though I honestly can't remember at the moment

julienw replied:
ok let's keep it here now then, but I'd like to know which bug @evhan55 found.
